databuilder/databuilder/models/type_metadata.py

```diff
@@ -0,0 +1,306 @@
+# SPDX-License-Identifier: Apache-2.0
+
+import abc
+from typing import Iterable, Iterator, Optional, Union
+
+from databuilder.models.graph_node import GraphNode
+from databuilder.models.graph_relationship import GraphRelationship
+from databuilder.models.graph_serializable import GraphSerializable
+from databuilder.models.table_metadata import DescriptionMetadata
+
+
+class TypeMetadata(abc.ABC, GraphSerializable):
+    TYPE_NODE_LABEL = 'Column_Subtype'
+    TYPE_RELATION_TYPE = 'COLUMN_SUBTYPE'
+    INVERSE_TYPE_RELATION_TYPE = 'COLUMN_SUBTYPE_OF'
+    TYPE_KIND = 'kind'
+    TYPE_NAME = 'name'
+    TYPE_DESCRIPTION = 'description'
+    TYPE_DATA_TYPE = 'data_type'
+    TYPE_MAP_KEY = 'map_key'
+    TYPE_MAP_VALUE = 'map_value'
+    TYPE_SORT_ORDER = 'sort_order'
+
+    @abc.abstractmethod
+    def __init__(self,
+                 start_label: Optional[str] = None,
+                 start_key: Optional[str] = None) -> None:
+        self.start_label = start_label
+        self.start_key = start_key
+
+        self._node_iter = self.create_node_iterator()
+        self._relation_iter = self.create_relation_iterator()
+
+    @abc.abstractmethod
+    def __str__(self) -> str:
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def is_terminal_type(self) -> bool:
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def create_node_iterator(self) -> Iterator[GraphNode]:
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def create_relation_iterator(self) -> Iterator[GraphRelationship]:
+        raise NotImplementedError
+
+    def create_next_node(self) -> Optional[GraphNode]:
+        try:
+            return next(self._node_iter)
+        except StopIteration:
+            return None
+
+    def create_next_relation(self) -> Optional[GraphRelationship]:
+        try:
+            return next(self._relation_iter)
+        except StopIteration:
+            return None
+
+
+class ArrayTypeMetadata(TypeMetadata):
+    def __init__(self,
+                 data_type: Union[TypeMetadata, str],
+                 start_label: Optional[str] = None,
+                 start_key: Optional[str] = None) -> None:
+        super(ArrayTypeMetadata, self).__init__(start_label, start_key)
+        self.data_type = data_type
+
+    def __str__(self) -> str:
+        return f"array<{self.data_type.__str__()}>"
+
+    def is_terminal_type(self) -> bool:
+        return not isinstance(self.data_type, TypeMetadata)
+
+    def _get_node_key(self, start_key: str) -> str:
+        return f"{start_key}/array"
+
+    def _get_node(self, start_key: str) -> GraphNode:
+        node = GraphNode(
+            key=self._get_node_key(start_key),
+            label=TypeMetadata.TYPE_NODE_LABEL,
+            attributes={
+                TypeMetadata.TYPE_KIND: 'array',
+                TypeMetadata.TYPE_DATA_TYPE: self.data_type.__str__()
+            }
+        )
+        return node
+
+    def _get_relation(self,
```

Comment:
i generally like to inline these into `create_node_iterator`. The separate function isn't necessary.